arangod/IResearch/IResearchDataStore.cpp

```diff
@@ -1632,107 +1585,86 @@ bool IResearchDataStore::exists(IResearchDataStore::Snapshot const& snapshot,
+    if (auto r = insertImpl(ctx); ADB_UNLIKELY(!r.ok())) {
       ctx.Abort();
-      return res;
+      return r;
     }
     TRI_IF_FAILURE("ArangoSearch::MisreportCreationInsertAsFailed") {
       return {TRI_ERROR_DEBUG};
     }
-    return res;
+    return {};
   }
-  auto* key = this;
-  // TODO FIXME find a better way to look up a ViewState
-  auto* ctx = basics::downCast<IResearchTrxState>(state.cookie(key));
-  if (!ctx) {
-    // '_dataStore' can be asynchronously modified
-    auto linkLock = _asyncSelf->lock();
-
-    if (!linkLock) {
-      // the current link is no longer valid (checked after ReadLock
-      // acquisition)
-      return {
-          TRI_ERROR_ARANGO_INDEX_HANDLE_BAD,
-          absl::StrCat("failed to lock ArangoSearch index '", index().id().id(),
-                       "' while inserting a document into it")};
-    }
 
-    TRI_ASSERT(_dataStore);  // must be valid if _asyncSelf->get() is valid
-    auto ptr = std::make_unique<IResearchTrxState>(
-        std::move(linkLock), *(_dataStore._writer), meta.hasNested());
-
-    ctx = ptr.get();
-    state.cookie(key, std::move(ptr));
-
-    if (!ctx) {
-      return {TRI_ERROR_INTERNAL,
-              absl::StrCat("failed to store state into a TransactionState for "
-                           "insert into ArangoSearch index '",
-                           index().id().id(), "', tid '", state.id().id(),
-                           "', revision '", documentId.id(), "'")};
-    }
-    state.addBeforeCommitCallback(&_beforeCommitCallback);
-    state.addAfterCommitCallback(&_afterCommitCallback);
+  TRI_IF_FAILURE("ArangoSearch::BlockInsertsWithoutIndexCreationHint") {
+    return {TRI_ERROR_DEBUG};
   }
 
+  auto* ctx = getContext(state);
+  if (ADB_UNLIKELY(!ctx)) {
+    return {TRI_ERROR_ARANGO_INDEX_HANDLE_BAD,
+            MakeMessage("While inserting document failed to lock", index(),
+                        state, documentId)};
+  }
   return insertImpl(ctx->_ctx);
 }
 
+template<typename FieldIteratorType, typename MetaType>
+void IResearchDataStore::recoveryInsert(uint64_t recoveryTick,
+                                        LocalDocumentId documentId,
+                                        velocypack::Slice doc,
+                                        MetaType const& meta) {
+  TRI_ASSERT(recoveryTickLow() < recoveryTick);
+  TRI_ASSERT(!isOutOfSync());
+  TRI_ASSERT(_recoveryRemoves);
+  try {
+    std::ignore = insertDocument<FieldIteratorType>(*this, _recoveryTrx, doc,
+                                                    documentId, meta);
+  } catch (std::bad_alloc const& e) {
+    throw;
+  } catch (...) {
+  }
+  if (!_recoveryTrx.FlushRequired()) {
+    return;
+  }
+  if (!_recoveryRemoves->empty()) {
+    _recoveryTrx.Remove(std::move(_recoveryRemoves));
+    _recoveryRemoves = makePrimaryKeysFilter(_hasNestedFields);
+  }
+  _recoveryTrx.Commit(recoveryTick);
+  // TODO(MBkkt) IndexWriter::Commit? Probably makes sense only if were removes
```

Comment:
Probably makes sense even without removes.
Because if we will make commit, even if recovery crashes, already committed part won't be executed second time